lib/irb/ruby-lex.rb

```diff
@@ -729,18 +730,27 @@ def check_string_literal(tokens)
           end
         end
       when :on_backtick
-        start_token << t
-        end_type << :on_tstring_end
+        if t.state.allbits?(Ripper::EXPR_BEG)
+          start_token << t
+          end_type << :on_tstring_end
+        end
       when :on_qwords_beg, :on_words_beg, :on_qsymbols_beg, :on_symbols_beg
         start_token << t
         end_type << :on_tstring_end
       when :on_heredoc_beg
-        start_token << t
-        end_type << :on_heredoc_end
+        pending_heredocs << t
+      end
+
+      if !pending_heredocs.empty? && t.tok.include?("\n")
+        pending_heredocs.reverse_each do |t|
+          start_token << t
+          end_type << :on_heredoc_end
+        end
+        pending_heredocs = []
       end
       i += 1
     end
-    start_token.last.nil? ? nil : start_token.last
+    pending_heredocs.first || start_token.last
```

Comment:
The tests seem to pass without `pending_heredocs.first ||`. Would you mind adding another test to cover this scenario?

Reply:
I added these test cases.
```ruby
['<<A;<<B', "<<A;<<B\n", "%W[\#{<<A;<<B", "%W[\#{<<A;<<B\n"]`
```
`process_string_literal` of these code should all be `<<A`.
without `pending_heredocs.first ||`, it  will be
```ruby
[nil, '<<A', '%W[', '<<A']
```

Reply:
👍 thx